WebGL2 3D Texture Support

mattdesl opened the pull request:
This is a first draft of 3D texture support, a WebGL2 feature.

``` js
var tex3d = new THREE.Texture3D( data, width, height, depth, format, type );
```

In FF Nightly and Chrome Canary, I'm getting some errors when using `gl.RGB` as an _internalFormat_. So, for now, this only supports RGB32F or RGBA32F as internal format.

One application of 3D textures is for lookup-table based color transforms. A 300es shader might look like this. Notice the `sampler3D` precision qualifier is necessary for this to compile, and it needs to be used as a RawShaderMaterial.

``` glsl
#version 300 es
precision mediump float;
precision highp sampler3D;

in vec2 vUv;
out vec4 fragColor;
uniform sampler2D tDiffuse;
uniform sampler3D tLookup;
const float lutSize = 32.0; // typically better as a uniform...

void main () {
  // Based on "GPU Gems 2 — Chapter 24. Using Lookup Tables to Accelerate Color Transformations"
  // More info and credits @ http://http.developer.nvidia.com/GPUGems2/gpugems2_chapter24.html
  vec4 rawColor = texture(tDiffuse, vUv);

  // Compute the 3D LUT lookup scale/offset factor
  vec3 scale = vec3((lutSize - 1.0) / lutSize);
  vec3 offset = vec3(1.0 / (2.0 * lutSize));

  // Apply 3D LUT color transform
  fragColor.rgb = texture(tLookup, scale * rawColor.rgb + offset).rgb;
  fragColor.a = rawColor.a;
}
```

More reading:
http://http.developer.nvidia.com/GPUGems2/gpugems2_chapter24.html

Related parser:
https://github.com/thibauts/parse-cube-lut/

mattdesl committed:
add check for webgl2 context

mattdesl committed:
add 3D texture support

mattdesl committed:
fix missing bracket

mattdesl committed:
fix wrapR setter

mattdesl committed:
more fixes to 3D texture bindings

mattdesl committed:
more fixes to get it working in Chrome

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2842,6 +2843,12 @@ THREE.WebGLRenderer = function ( parameters ) {
 
+			if ( texture instanceof THREE.Texture3D ) {
+				if ( !_isWebGL2 ) throw new Error('3D Textures are only supported in WebGL2 contexts.');
+				if ( !texture.wrapR ) console.warn('Texture3D is missing a .texture.wrapR option!');
```
I don't think this is needed? `Texture3D` has `wrapR`.

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2885,6 +2892,17 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 	}
 
+	function getTextureTarget ( texture ) {
```
Maybe `getTextureType` is a better name?

mattdesl replied:
Hmm the only thing is that `texture.type` refers to unsigned byte / float / etc. Whereas "target" is used in GL spec to mean `gl.TEXTURE_2D` vs `gl.TEXTURE_3D` etc.

mrdoob replied:
I see I see.

mrdoob commented:
If you have a chance, it would be nice to have an example too.

mattdesl committed:
remove unnecessary lines; the WebGL2 check is already happening in getTextureTarget method

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2885,6 +2890,17 @@ THREE.WebGLRenderer = function ( parameters ) {
+			textureTarget = _gl.TEXTURE_3D;
+		}
+		return textureTarget;
+	}
```
Another simplification...

```javascript
function getTextureTarget ( texture ) {
	if ( texture instanceof THREE.Texture3D ) {
		if ( !_isWebGL2 ) {
			throw new Error('3D Textures are only supported in WebGL2 contexts.');
		}
		return _gl.TEXTURE_3D;
	}
	return _gl.TEXTURE_2D;
}
```

bhouston commented:
Would love to see an example as well as @mrdoob suggested.  PRs are easier to get accepted if there is an example, and having webgl2_*.html examples is a really good idea.

mattdesl committed:
Merge branch 'dev' into webgl2-3d-texture

mattdesl committed:
add example; simplify function; expose isWebGL2 parameter on Renderer

mattdesl commented:
Updated PR, should be good to go now.

- Simplified `getTextureTarget` in WebGLRenderer
- Added `examples/webgl2_3d_texture.html` and `examples/lut/ColorTransform.CUBE`
- Exposed `renderer.isWebGL2`, a boolean which is true when the context is a WebGL2 instance

mattdesl committed:
fix title

mattdesl commented:
Live WebGL2 example:
https://dl.dropboxusercontent.com/u/74783296/tests/webgl2-3d-texture/webgl2_3d_texture.html

bhouston commented:
This is so awesome. :)  Although I want to get 3D LUTs working without WebGL 2.0.

bhouston commented:
I support getting this merged. :)

bhouston commented:
@mattdesl can you add your example to examples/files.js?

bhouston commented:
@mattdesl One other question, I assume this is a linear LUTs so it can not handle HDR colors?

WestLangley commented:
@mattdesl 

This may be more convenient for you.

http://ci.threejs.org/api/pullrequests/8124/examples/webgl2_3d_texture.html

You just need to include `webgl2_3d_texture.html` in `files.js` in this PR and it will appear in the three.js examples at http://threejsworker.com/.

mattdesl committed:
add files

mattdesl commented:
@WestLangley Woah, that's awesome!

@bhouston To get it working on my linear-space images I had to correct the values like this:

```js
const gamma = 2.2;
const gammaCorrect = Boolean(opt.sRGB);
const float32 = new Float32Array(length);
for (let k = 0, i = 0; i < lutData.length; i++) {
  for (let j = 0; j < 3; j++) {
    let component = lutData[i][j];
    if (gammaCorrect) {
      component = Math.pow(component, 2.2);
    }
    float32[k++] = component;
  }
}
```

Then I did `pow(rawColor.rgb, vec3(1.0 / GAMMA))`  before passing the color to the 3D table lookup.

bhouston commented:
@mattdesl that is just gamma correction, not HDR (which means it could go from 0 to a very high number.)  That said, we should still merge this and then look to add non-linear LUTs.

If you wanted to handle gamma correction faster you could likely bake it into your LUT prior to GPU upload.  Maybe we should create a LUT JavaScript class that can do transforms like that, or handle HDR LUTs and conversion to 3D or 2D textures?

mattdesl commented:
The LUT data for the example file is bounded to `0 ... 1`. Input colors outside that range may clamp / wrap depending on the texture parameters. Is that what you are referring to?

I'm not sure how to export a different domain from Photoshop, but the format supports a variable domain (other than 0 to 1).

Also, it would be great to have a 3D-to-2D tool. I've already done a bit of work in this area – see [glsl-lut](https://www.npmjs.com/package/glsl-lut) – but nothing to generate the 2D LUT itself.

bhouston commented:
I am not a LUT expert.  Should one apply LUTs before tone mapping to the 0 to 1 range or after?  What is standard practice?  I know that Sketchfab applies color correction (which LUTs can represent) after tone mapping.  I will research this.

Best stuff on the 2d front. :). Thanks for your amazing contributions.

bhouston commented:
It seems unreal engine also applies the LUT after tone mapping thus no HDR support is required.  Nice.

rc1 commented:
Hi just stumbled into this issue...

Example code [here](https://gist.github.com/rc1/433177bd0c2ce38fcccd#file-lutstringtotexture-js) which loads a `.cube` text file (generated from the free version of DaVinci Resolve) and creates a texture like attached. The fragment shader then [samples it as a 3D texture](https://gist.github.com/rc1/433177bd0c2ce38fcccd#file-lut-frag). 

![screen shot 2016-03-24 at 10 45 05](https://cloud.githubusercontent.com/assets/166915/14014966/c6433ff8-f1ad-11e5-883e-05a9a85bb9dd.png)

mattdesl committed:
Merge branch 'dev' into webgl2-3d-texture

mrdoob commented:
@mattdesl Any chance you could give another go at implementing this on top of the latest `dev`?

matgr1 commented:
I needed this, so I did it a little while ago using @mattdesl's changes (and I just merged in the latest changes from dev)... if you want I'll check your contributing guidelines and put together a pull request?

mrdoob commented:
Okey... sorry for the huge delay.

I was hesitant to merge this because I wasn't sure I wanted to go down the path of making `WebGLRenderer` support `webgl2` too...

Say hello to `WebGL2Renderer`!  https://github.com/mrdoob/three.js/commit/2ff9d410753b72a5e43b211dc3be26f0f0ab8a0e 👋

A nice thing about `WebGL2Renderer` is that it gives us a chance to clean things up, starting with only supporting `BufferGeometry` ✌️

medihack commented:
Any updates on having 3D textures in three.js? I would like to use three.js with some medical imaging stuff and it would help volume rendering a lot.

medihack commented:
One thing I don't like that much is that you have to use a `RawShaderMaterial`. Wouldn't it be nice in `WebGLProgram` to check a flag of the `ShaderMaterial` if it is a WebGL 2.0 shader and then add "#version 300 es" upfront? And maybe set the flag like this `var material = new THREE.ShaderMaterial({...., webgl2: true});`
